ADR/0032-decoupling-deployment.md

```diff
@@ -0,0 +1,201 @@
+# 32. Decoupling Deployment
+
+Date: 2023-11-17
+
+## Status
+
+Draft
+
+## Authors
+
+- Ralph Bean
+
+## Context
+
+Since the beginning of our project, we've had an emphasis on providing an integrated experience for
+the user, automating all steps through build, test, deployment, and release to higher environments.
+
+Some challenges:
+
+- Our controllers' APIs are highly coupled. For example, it isn't feasible to try and use
+  [integration-service] today without [application-service] and [gitops-service]. This creates
+  a high barrier to engage with some of the useful subsystems we've built. A user (and potential
+  contributor) has to adopt the whole thing, or nothing.
+- Today, we can generate simple but often incorrect deployment manifests for users' applications,
+  which we deploy via the application gitops repo. The user cannot interact with or influence these
+  resources directly; they have to do so through the Application/Component API. If we permit them to
+  directly provide deployment resources through that API, the API will become a leaky abstraction.
+  If don't permit them to provide their own resources, and instead try to handle every application
+  configuration case in the HAS generation code, we will struggle to keep up with users' demands for
+  the many and varied kinds of application deployments.
+
+See also [RHTAP-1873](https://issues.redhat.com/browse/RHTAP-1873).
+
+## Decision
+
+We are going to decouple the deployment from the build, test, and release portions of our system.
+
+**Deployment**:
+
+- The [Environment], [SnapshotEnvironmentBinding], and [GitOpsDeploymentManagedEnvironment]
+  resources will be deprecated and eventually decomissioned.
+- The [application-service] will stop generating GitOps repo content.
+- HAC will stop rendering [Environments] and their status.
+- If a user wants to make use of deployment capabilities, we will promote the usage of [renovatebot]
+  to propagate _released_ images to their self-managed gitops repo as pull requests.
+
+**Test and promotion**:
+
+- We will retain the [DeploymentTargetClaim], [DeploymentTarget], and [DeploymentTargetClass] APIs.
+  Collectively, call these the "DTC" APIs. These should have no dependence on other AppStudio APIs.
+  Only [integration-service] should depend on them. The controllers that support the DTC API should
+  be organized into their own project: "the [deployment-target-operator]", or something like that.
```

Comment:
Do these CRD names still make sense? If the system (i.e. AppStudio) is no longer deploying Applications on environments, then I feel like these should be abstracted further to indicate the intended functionality. Namely that there is some system of provisioning environments. The fact that these are intended for deployment by integration service is no longer relevant to the underlying processes given that we want to isolate it from AppStudio.

Reply:
Ack - renaming them is an option. I think they still make sense on a certain level, but the broadness of their potential usage means we could revisit that for sure.